src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs

```diff
@@ -0,0 +1,230 @@
+            EnsureArg.IsNotNull(action, nameof(action));
+            EnsureArg.IsTrue(_modelInfoProvider.IsKnownResource(resourceType), nameof(resourceType), x => GenerateTypeErrorMessage(x, resourceType));
+
+            ListedRestComponent listedRestComponent = _statement.Rest.Server();
+            ListedResourceComponent resourceComponent = listedRestComponent.Resource.SingleOrDefault(x => string.Equals(x.Type, resourceType, StringComparison.OrdinalIgnoreCase));
+
+            if (resourceComponent == null)
+            {
+                resourceComponent = new ListedResourceComponent
+                {
+                    Type = resourceType,
+                };
+                listedRestComponent.Resource.Add(resourceComponent);
+            }
+
+            action(resourceComponent);
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder TryAddRestInteraction(string resourceType, string interaction)
+        {
+            EnsureArg.IsNotNullOrEmpty(resourceType, nameof(resourceType));
+            EnsureArg.IsNotNullOrEmpty(interaction, nameof(interaction));
+            EnsureArg.IsTrue(_modelInfoProvider.IsKnownResource(resourceType), nameof(resourceType), x => GenerateTypeErrorMessage(x, resourceType));
+
+            UpdateRestResourceComponent(resourceType, c =>
+            {
+                c.Interaction.Add(new ResourceInteractionComponent
+                {
+                    Code = interaction,
+                });
+            });
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder TryAddRestInteraction(string systemInteraction)
+        {
+            EnsureArg.IsNotNullOrEmpty(systemInteraction, nameof(systemInteraction));
+
+            _statement.Rest.Server().Interaction.Add(new ResourceInteractionComponent { Code = systemInteraction });
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder TryAddSearchParams(string resourceType, IEnumerable<SearchParamComponent> searchParameters)
+        {
+            EnsureArg.IsNotNullOrEmpty(resourceType, nameof(resourceType));
+            EnsureArg.IsNotNull(searchParameters, nameof(searchParameters));
+            EnsureArg.IsTrue(_modelInfoProvider.IsKnownResource(resourceType), nameof(resourceType), x => GenerateTypeErrorMessage(x, resourceType));
+
+            UpdateRestResourceComponent(resourceType, c =>
+            {
+                foreach (SearchParamComponent searchParam in searchParameters)
+                {
+                    c.SearchParam.Add(searchParam);
+                }
+            });
+
+            return this;
+        }
+
+        public ICapabilityStatementBuilder AddDefaultResourceInteractions()
+        {
+            foreach (var resource in _modelInfoProvider.GetResourceTypeNames())
+            {
+                if (string.Equals(resource, KnownResourceTypes.Parameters, StringComparison.Ordinal))
```

Comment:
nit: maybe a bit of comment here will help people later? Why we are skipping certain types and what not?